### PR TITLE
feat(weighting): add decay function calculator for gravity-based weighting

### PR DESCRIPTION
**Pull request by kaligrafy**
Add DecayFunctionCalculator service implementing distance and time decay functions for gravity-based accessibility and weighting models. Supports five decay function types:
- Power: f(x) = x^(-beta)
- Exponential: f(x) = exp(-beta * x)
- Gamma: f(x) = a * x^(-b) * exp(-c * x)
- Combined: f(x) = x^(-beta1) * exp(-beta2 * x)
- Logistic: f(x) = 1 / (1 + exp(beta * (x - x0)))

The calculator accepts DecayInputValue objects containing both distance and time values, with xType parameter determining which value to use. Includes parameter validation, threshold handling for minimum values, and comprehensive test coverage.

Add types.ts with type definitions for decay function parameters and input values. Test suite includes validation tests, calculation tests for all function types, and edge case handling.

## Summary by CodeRabbit

* **New Features**
  * Adds configurable decay models (power, exponential, gamma, combined, logistic), minimum distance/time thresholds, input validation, and clear error responses for invalid configurations.
* **Tests**
  * Adds comprehensive tests covering calculations, parameter validation, boundary and edge cases, error handling, cross-type scenarios, and extreme values.

### CODE REVIEW

File: packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+export class DecayFunctionCalculator {
+    /**
+     * Get the input value (distance or time) from the DecayInputValue object based on xType
+     *
+     * @param inputValue Value object containing both distance and time
+     * @param xType Type of the input value (distance or time)
+     * @returns The distance or time value based on xType
+     */
+    static getInputValue(inputValue: DecayInputValue, xType: DecayInputValueType): number {
+        return xType === 'distance' ? inputValue.distance : inputValue.time;
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts]: fixed

File: packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+            throw new TrError('Input value must be a finite non-negative number', 'DECAYFUNC003', 'WeightingError');
+        }
+
+        if (!Number.isFinite(inputValue.distance) || inputValue.distance < 0) {
+            throw new TrError('Distance value must be a finite non-negative number', 'DECAYFUNC017', 'WeightingError');
+        }
+
+        if (!Number.isFinite(inputValue.time) || inputValue.time < 0) {
+            throw new TrError('Time value must be a finite non-negative number', 'DECAYFUNC018', 'WeightingError');
+        }
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts]: @coderabbitai Is this fixed?

**Comment by kaligrafy** [comment]
@coderabbitai please re-review. Thanks!

File: packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts [hunk 3]
```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright 2024, Polytechnique Montreal and contributors
```
Comment (tahini) [inline packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts]: 2025

File: packages/transition-backend/src/services/weighting/types.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    b: number;
+    c: number;
+}
+
+/**
+ * Parameters for combined decay function: f(x) = x^(-beta1) * exp(-beta2 * x)
+ *
+ * Constraints:
+ * - beta1 must be > 0 for power-law component (typical range: 0.5 to 2.0)
+ * - beta2 must be >= 0 for exponential component (beta2 = 0 results in pure power decay)
```
Comment (tahini) [inline packages/transition-backend/src/services/weighting/types.ts]: Pourquoi supporter beta2 === 0 au lieu de juste prendre un decay function directement?

File: packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+ * - Logistic: f(x) = 1 / (1 + exp(beta * (x - x0)))
+ *
+ * Where x can be either distance (in meters) or time (in seconds).
+ */
+export class DecayFunctionCalculator {
+    /**
+     * Get the input value (distance or time) from the DecayInputValue object based on xType
+     *
+     * @param inputValue Value object containing both distance and time
+     * @param xType Type of the input value (distance or time)
```
Comment (tahini) [inline packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts]: inputValueType?

**Review by tahini — COMMENTED** [review]
Quelques questionnements sur la structure. À toi de voir selon les réponses s'il y a des changements à apporter. Mais dans tous les cas, il faut mettre la visibilité des fonctions au minimum nécessaire pour le besoin. Les fonctions comme `getInputValue` seront testées indirectement en testant la fonction de calcul.

File: packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts (hunk 5 above)
Comment (kaligrafy) [reply packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts]: replaced

File: packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+         * are indeed not possible.
+         * So we need to put a minimum threshold in the input values for some functions
+         * that would be undefined dividing by 0 or return numbers near infinity.
+         * For travel time, we will use a minimum of 60 seconds.
+         * For distance, we will use a minimum of 100 meters.
+         * This is a compromise between the accuracy of the model and the practicality of the data.
+         * However, using the exponential decay function, the result would be close to 1 for very small values,
+         * so this would be the best function to use if very small values are expected.
+         */
+        const adjustedX = this.getAdjustedValue(inputValue, xType);
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts]: Il va falloir tester de toute façon avec des tests fonctionnels et rouler des simulations. On pourra ajuster plus tard.

File: packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+     * @param xType Type of the input value (distance or time) - determines which value from inputValue is used
+     * @param parameters Decay function parameters
+     * @returns Decay function value (friction factor)
+     * @throws TrError if parameters are invalid
+     */
+    static calculateDecay(
+        inputValue: DecayInputValue,
+        xType: DecayInputValueType,
+        parameters: DecayFunctionParameters
+    ): number {
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts]: Je pense que oui, en tout cas en attendant d'avoir fait des tests avec plusieurs simulations. On pourra toujours y revenir plus tard.

[the remark this answers is not in the document]
Comment (tahini) [reply packages/transition-backend/src/services/weighting/DecayFunctionCalculator.ts]: > Je pense que oui, en tout cas en attendant d'avoir fait des tests avec plusieurs simulations. On pourra toujours y revenir plus tard.

Tu penses que oui à quel élément de mon commentaire? En tout, je ne suis pas convaincue ici, je pense qu'on risque d,appeler une même méthode plusieurs fois, en tout cas pour la simulation, et qu'une approche par classe/objet au lieu d'un switch case serait plus efficace. Mais je ne bloquerais pas sur ça, en l'utilisant, on verra ben.
